### PR TITLE
opencomposite: init at unstable-2023-07-02

### PR DESCRIPTION
**Pull request by Scrumplex**
###### Description of changes

[OpenComposite](https://gitlab.com/znixian/OpenOVR) is a free and open source implementation of Valve's OpenVR on top of OpenXR.

Despite its name, OpenVR's "reference" implementation (SteamVR) is proprietary.

I also included a helper package (`opencomposite-helper`), that sets the appropriate environment variables to make it work with Proton games on Steam. I am not sure if we want to keep this in nixpkgs though.

I personally tested this with the game Hot Dogs, Horseshoes & Hand Grenades (H3VR) on a Valve Index using my NixOS module proposed in #245005.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/opencomposite/default.nix
Comment (drupol) [inline pkgs/applications/graphics/opencomposite/default.nix:28]: Each patch in nixpkgs applied to the upstream source should be documented. Out-of-tree patches, fetched using `fetchpatch`, are preferred.

In each patch comment, please explain the purpose of the patch and link to the relevant upstream issue if possible. If the patch has been merged upstream but is not yet part of the released version, please note the version number or date in the comment such that a future maintainer updating the nix expression will know whether the patch has been incorporated upstream and can thus be removed from nixpkgs.

Furthermore, please use a _stable_ URL for the patch. Rather than, for example, linking to a GitHub pull request of the form `https://github.com/owner/repo/pull/pr_number.patch`, which would change every time a commit is added or the PR is force-pushed, link to a specific commit patch in the form `https://github.com/owner/repo/commit/sha.patch`.

Here are two good examples of patch comments:

```nix
mkDerivation {
  …

  patches = [
    # Ensure RStudio compiles against R 4.0.0.
    # Should be removed next 1.2.X RStudio update or possibly 1.3.X.
    (fetchpatch {
      url = "https://github.com/rstudio/rstudio/commit/3fb2397c2f208bb8ace0bbaf269481ccb96b5b20.patch";
      sha256 = "0qpgjy6aash0fc0xbns42cwpj3nsw49nkbzwyq8az01xwg81g0f3";
    })
  ];
}
```

```nix
mkDerivation {
  …

  patches = [
    # Allow building with bison 3.7
    # PR at https://github.com/GoldenCheetah/GoldenCheetah/pull/3590
    (fetchpatch {
      url = "https://github.com/GoldenCheetah/GoldenCheetah/commit/e1f42f8b3340eb4695ad73be764332e75b7bce90.patch";
      sha256 = "1h0y9vfji5jngqcpzxna5nnawxs77i1lrj44w8a72j0ah0sznivb";
    })
  ];
}
```

File: pkgs/applications/graphics/opencomposite/default.nix
Comment (drupol) [inline pkgs/applications/graphics/opencomposite/default.nix:52]: Could you please add the `meta.mainProgram` attribute?

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, let me know if this is clear enough.

File: pkgs/applications/graphics/opencomposite/default.nix
Comment (Scrumplex) [reply pkgs/applications/graphics/opencomposite/default.nix:28]: I sadly don't have an upstream patch for this. I could try to improve the detection of system libraries in the upstream project before we merge this.

I am going to add a comment about the patch for now.

File: pkgs/applications/graphics/opencomposite/default.nix
Comment (Scrumplex) [reply pkgs/applications/graphics/opencomposite/default.nix:52]: There is no mainProgram for the main `opencomposite` package

Comment (Scrumplex) [reply pkgs/applications/graphics/opencomposite/default.nix]: I just realized, that the path didn't quite make sense. I have moved this to `pkgs/development/libraries` instead, as it's really just an alternative implementation of `openvr`

**Review by drupol — APPROVED** [review]
LGTM

**Pull request merged**
